src/interface.jl

```diff
@@ -2,27 +2,90 @@
 
 
 """
-    hasdensity(d)::Bool
+    ismeasure(object)::Bool
 
-Return `true` if `d` is compatible with the `DensityInterface` interface.
+Return `true` if `object` is a measure.
 
-`hasdensity(d) == true` implies that `d` is either a density itself or has an
-associated density, e.g. a probability density function or a Radon–Nikodym
+!!ToDo: More text!!
+
+Defaults to `false`. For types that are measures (not densities), define
+
+```julia
+@inline ismeasure(::MyMeasureType) = true
+```
+
+The above will automatically provide `hasdensity(::MyMeasureType) == true`.
+
+See also [`hasdensity`](@ref) and [`isdensity`](@ref).
+"""
+function ismeasure end
+export ismeasure
+
+@inline ismeasure(object) = false
+
+
+"""
+    hasdensity(object)::Bool
+
+Return `true` if `object` is compatible with the `DensityInterface` interface.
+
+`hasdensity(object) == true` implies that `object` is either a density itself or
+has an associated density, e.g. a probability density function or a Radon–Nikodym
 derivative with respect to an implicit base measure. It also implies that the
 value of that density at given points can be calculated via
 [`logdensityof`](@ref) and [`densityof`](@ref).
+
+Defaults to `ismeasure(object)`. For types that are densities, not measures,
+define
+
+```julia
+@inline hasdensity(::MyDensityType) = true
 ```
+
+See also [`ismeasure`](@ref) and [`isdensity`](@ref).
 """
 function hasdensity end
 export hasdensity
 
-@inline hasdensity(::Any) = false
+@inline hasdensity(object) = ismeasure(object)
 
 function check_hasdensity(d)
     hasdensity(d) || throw(ArgumentError("Object of type $(typeof(d)) is not compatible with DensityInterface"))
 end
 
 
+"""
+    isdensity(object) = hasdensity(object) && !ismeasure(object)
+
+Return `true` if `object` is a density. Defaults to
+`hasdensity(object) && !ismeasure(object)`.
+
+!!ToDo: More text!!
+
+`isdensity` should *not* be specialized, specialize [`ismeasure`](@ref) instead.
+
+See also [`hasdensity`](@ref).
+"""
+function isdensity end
+export isdensity
+
+@inline isdensity(object) = hasdensity(object) && !ismeasure(object)
+
+
+"""
+    basemeasure(m)
+
+Return the base measure of measure `m`.
+
+`logdensityof(m, x)` and `densityof(m, x)` return the log/non-log
+value of the Radon–Nikodym derivative of `m` and `basemeasure(m)`.
+
+!!ToDo: More text!!
+"""
+function basemeasure end
```

Comment:
What's the reason for including it in DensityInterface and not keeping it in MeasureTheory? Densities without `basemeasure`, such as user-defined densities or dynamic probabilistic models with implicit reference measure can't/won't implement this function. It seems the set of densities that packages that handle reference measures explicitly will always be a strict subset of all supported densities.

Mainly I wonder which packages would actually implement this function - I assume mainly MeasureTheory and Distributions? In this case I don't see a clear advantage over the current design - MeasureTheory could just own `basemeasure` and define it for `Distributions.Distribution`s, as currently done.

In fact this would have the following advantage: One would not have to define a set of base measures in DensityInterface that can be shared between Distributions and MeasureTheory. Without these it seems the `basemeasure` itself would not be sufficient for interoperability of the packages. However, arguably it seems a bit simpler if MeasureTheory defines measures and one does not have to determine which ones should be moved to DensityInterface.

Reply:
I think by including `basemeasure` in `DensityInterface` we can have avoid having a direct dependency via Distributions and MeasureTheory/MeasureBase. That's one important potential of lightweight central interfaces, after all: Allowing composability without dependency or `@requires`.

We could, of course, have another lightweight package, secondary to DensityInterface, to define `basemeasre`. If we use a type-based `densityinfo`, we could maybe even move everything with `measure` in it's name into such a secondary lightweight package - I'm not sure about that, though.

I'd suggest for now let's prototype all the "super-lightweight" aspects here, so can play and test with DensityInterface, Distributions and MeasureBase/MeasureTheory. Once we have converged on a design it will probably be clearer what should go into DensityInterface itself.

Reply:
My suggestion would be to keep everything regarding `basemeasure`, measures, etc. in MeasureTheory. No changes in Distributions at all. MeasureTheory defines measures and `basemeasure` for Distributions` distributions. This would mean no type piracy and no issues with measures in DensityInterface/another package.

I don't see any advantage in extracting `basemeasure` and measures from MeasureTheory as long as there are no other packages than Distributions and MeasureTheory that would make use of it. Of course, one can always reevaluate once other packages would like to work with `basemeasure` etc. but at this point I think it creates problems without solving an issue.

Reply:
I'm kind of on the fence about this myself. The tradeoff as I see it is

## `MeasureBase.basemeasure`

- More control and responsibility to MeasureTheory
- Probably not too bad to get all of Distributions covered
- But Distributions users wanting an easy `basemeasure` will need to depend on MeasureBase (easily avoided as you describe)
- Maybe the biggest potential issue... Say someone makes a new package extending Distributions. To use this from MeasureTheory would require implementing `basemeasure`. But where should that go? It seems this would lead to either lots of replicated code, or a glue package for each such case.

## `DensityInterface.basemeasure`

- A step toward allowing singular mixtures in Distributions (otherwise impossible)
- Would lead to trouble if some methods are left missing or incorrect
- Requires Distributions having access to some non-distributional measures (Lebesgue, CountingMeasure, etc)
- Would define things in one place (DensityInterface or Distributions) and delegate all the tools for working with them to another (MeasureBase). That's not inherently bad, but seems a little odd. We'd also have to be careful to set things up in a way that doesn't lead us to type piracy later